src/Datadog.AutoInstrumentation.NativeLoader/dynamic_instance.cpp

```diff
@@ -0,0 +1,167 @@
+#include "dynamic_instance.h"
+
+#include "guid.h"
+#include "logging.h"
+#include "pal.h"
+
+namespace datadog::shared::nativeloader
+{
+
+    // ************************************************************************
+
+    //
+    // private
+    //
+
+    HRESULT DynamicInstanceImpl::EnsureDynamicLibraryIsLoaded()
+    {
+        if (!m_loaded)
+        {
+            m_instance = LoadDynamicLibrary(m_filepath);
+            m_loaded = true;
+        }
+
+        return m_instance != nullptr ? S_OK : E_FAIL;
+    }
+
+    HRESULT DynamicInstanceImpl::DllGetClassObject(REFIID riid, LPVOID* ppv)
+    {
+        // Check if the library is loaded
+        if (FAILED(EnsureDynamicLibraryIsLoaded()))
+        {
+            return E_FAIL;
+        }
+
+        // Check if the function pointer needs to be loaded
+        if (m_getClassObjectPtr == nullptr)
+        {
+            m_getClassObjectPtr =
+                static_cast<dllGetClassObjectPtr>(GetExternalFunction(m_instance, "DllGetClassObject"));
+        }
+
+        // If we have the function pointer we call the function
+        if (m_getClassObjectPtr != nullptr)
+        {
+            return m_getClassObjectPtr(m_clsid, riid, ppv);
+        }
+
+        // The function cannot be loaded.
+        return E_FAIL;
+    }
+
+    //
+    // public
+    //
+
+    DynamicInstanceImpl::DynamicInstanceImpl(std::string filePath, std::string clsid)
+    {
+        m_filepath = filePath;
+        m_clsid_string = clsid;
+        m_clsid = guid_parse::make_guid(clsid);
+        m_loaded = false;
+        m_instance = nullptr;
+        m_getClassObjectPtr = nullptr;
+        m_canUnloadNow = nullptr;
+        m_classFactory = nullptr;
+        m_corProfilerCallback = nullptr;
+    }
+
+    DynamicInstanceImpl::~DynamicInstanceImpl()
+    {
+        m_corProfilerCallback = nullptr;
+        m_classFactory = nullptr;
+        m_canUnloadNow = nullptr;
+        m_getClassObjectPtr = nullptr;
+        m_loaded = false;
+
+        if (m_instance != nullptr)
+        {
+            if (!FreeDynamicLibrary(m_instance))
+            {
+                Warn("DynamicInstanceImpl::~DynamicInstanceImpl: Error unloading: ", m_filepath, " dynamic library.");
+            }
+            m_instance = nullptr;
+        }
+    }
+
+    HRESULT DynamicInstanceImpl::LoadClassFactory(REFIID riid)
+    {
+        LPVOID ppv;
+        HRESULT res = DllGetClassObject(riid, &ppv);
+        if (SUCCEEDED(res))
+        {
+            m_classFactory = static_cast<IClassFactory*>(ppv);
+        }
+        else
+        {
+            Warn("DynamicInstanceImpl::LoadClassFactory: Error getting IClassFactory from: ", m_filepath);
+        }
+
+        Debug("DynamicInstanceImpl::LoadClassFactory: ", res);
+        return res;
+    }
+
+    HRESULT DynamicInstanceImpl::LoadInstance(IUnknown* pUnkOuter, REFIID riid)
+    {
```

Comment:
Is not used but the signature of this method meant to be similar to the class factory `CreateInstance` method without the returning `ppvObject` due we store that one in `m_corProfilerCallback` of the same `DynamicInstanceImpl` instance

Reply:
It looks like a bug to the reader though. I'd prefer if this would be either piped through or removed from the signature (can always add it as needed). After all the id is a parameter on the normal factory method, although we can only create instances of that one class. Otherwise, it looks to a reader like something is forgotten here. However, I do not feel strongly about this. I'll add a follow up tag, so that this gets noticed, but feel free to skip.

Reply:
❗ [follow up in future PR]